### PR TITLE
feat: add in reporting functions

### PR DESCRIPTION
**Pull request by iamnewton**
**Description**

Add in support for buyer and seller reporting functions.

**Screenshots (optional)**

N/A

**Additional context (optional)**

N/A

**Checklist**
_Have you considered all of the following?_

Please confirm by checking off each item to ensure you've considered it:

- [x] Did you create or update tests? 
- [x] Did you create or update any documentation?

---

Fixes: #56

### CODE REVIEW

File: src/render/index.js [hunk 1]
```diff
@@ -63,9 +63,13 @@ export default async function renderAd (selector, token, debug = false) {
...
 	debug && echo.info('sending reports to the seller');
-	const sellersReport = getSellerReport(winner.conf, winner);
-	debug && echo.info('sending reports to the buyer');
-	getBuyerReport(winner.conf, winner, sellersReport);
+	const sellersReport = await getSellerReport(winner.conf, winner);
+	if (sellersReport) {
+		debug && echo.info('sending reports to the buyer', sellersReport);
+		await getBuyerReport(winner.conf, winner, sellersReport);
+	} else {
+		debug && echo.error('no report was sent to the buyer');
```
Comment (brodrigu) [inline src/render/index.js:71]: if no sellers report we should still call buyers report, but pass null in `sellersReport` so the buyer knows there might have been an issue.

File: src/reporting/utils.js [hunk 2]
```diff
@@ -70,14 +69,18 @@ export const getSellerReport = async (conf, results) => {
...
  * @param {object} conf - an auction configuration object
  * @param {object} results - the results of the auction
+ * @param {object} report - the report object from the sellers report
  * @return {void} has a side effect of generating a report for the buyer
  */
 export const getBuyerReport = async (conf, results, report) => {
-	const { report_win } = await import(conf.bidding_logic_url);
+	if (!report) {
+		return null;
+	}
```
Comment (brodrigu) [inline src/reporting/utils.js]: buyer should still get a chance to send a report even if seller messed up their report

File: src/render/index.js (hunk 1 above)
Comment (iamnewton) [reply src/render/index.js]: k, fixed

File: src/reporting/utils.js (hunk 2 above)
Comment (iamnewton) [reply src/reporting/utils.js]: k, fixed

File: src/render/reporting.js [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+export const getSellerReport = async (conf, results) => {
+	const { report_result } = await import(conf.decision_logic_url);
+
+	// check if there is even a function
+	if (!report_result && typeof report_result !== 'function') {
+		return null;
+	}
+
+	// generate a report by providing all of the necessary information
+	return report_result(conf, {
```
Comment (brodrigu) [inline src/render/reporting.js:56]: ideally we would have an `await` and a `try`/`catch` here as the report function will require an async call to the sellers server to report.

File: src/render/reporting.js [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+export const getBuyerReport = async (conf, results, report) => {
+	const { report_win } = await import(results.bid.bidding_logic_url);
+
+	// check if there is even a function
+	if (!report_win && typeof report_win !== 'function') {
+		return null;
+	}
+
+	// generate a report by providing all of the necessary information
+	return report_win(conf?.auction_signals, conf?.per_buyer_signals?.[results.bid.owner], report, {
```
Comment (brodrigu) [inline src/render/reporting.js:84]: await + try/catch

File: src/render/reporting.js (hunk 3 above)
Comment (brodrigu) [reply src/render/reporting.js:56]: ok I get why you don't need the await (since you are returning a promise here) but you'll need the try/catch

Comment (brodrigu) [reply src/render/reporting.js:56]: and since you'll be adding a try/catch you'll need the await, ha ha

Comment (iamnewton) [reply src/render/reporting.js:59]: This works the same as the bidding utility function. It's a dynamic import so no need for that. If you look at the demo it works and calls the function and returns data in the console.

**Comment by brodrigu** [comment]
please also add try/catch to the bidding and scoring function calls. catching a bid error should just ignore that bid, catching a scoring error should ignore that bids score and continue scoring the next one.

File: src/render/reporting.js (hunk 3 above)
Comment (iamnewton) [reply src/render/reporting.js]: fixed

File: src/auction/utils.js [hunk 5]
```diff
@@ -141,6 +155,7 @@ const getTrustedSignals = async (url, keys) => {
 			return response.json();
 		})
 		.catch(error => {
+			// @TODO, probably want to return `null` here too
```
Comment (brodrigu) [inline src/auction/utils.js:158]: agreed. does not need to be in this PR though

File: src/render/index.js [hunk 6]
```diff
@@ -51,10 +56,17 @@ export default async function renderAd (selector, token, debug = false) {
 	debug && echo.info('checking that ad iframe actually rendered');
 	const ad = getTarget(`#fledge-auction-${token}`);
 	debug && echo.log('ads target:', ad);
-	if (!ad) {
+	if (!ad && hasRendered(ad)) {
```
Comment (brodrigu) [inline src/render/index.js:59]: should this be `if( !ad || !hasRendered(ad))`? 

I'm not sure how we would not have an ad, and have it rendered.

File: src/render/reporting.js [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+ * @author Newton <cnewton@magnite.com>
+ * @param {object} conf - an auction configuration object
+ * @param {object} results - the results of the auction
+ * @return {object} an object of data to pass back to the buyers report
+ */
+export const getSellerReport = async (conf, results) => {
+	const { report_result } = await import(conf.decision_logic_url);
+
+	// check if there is even a function
+	if (!report_result && typeof report_result !== 'function') {
```
Comment (brodrigu) [inline src/render/reporting.js:55]:
```suggestion
	if (!report_result || typeof report_result !== 'function') {
```

File: src/render/reporting.js [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+ * @param {object} conf - an auction configuration object
+ * @param {object} results - the results of the auction
+ * @param {object} report - the report object from the sellers report
+ * @return {void} has a side effect of generating a report for the buyer
+ */
+export const getBuyerReport = async (conf, results, report) => {
+	const { report_win } = await import(results.bid.bidding_logic_url);
+
+	// check if there is even a function
+	if (!report_win && typeof report_win !== 'function') {
```
Comment (brodrigu) [inline src/render/reporting.js:88]:
```suggestion
	if (!report_win || typeof report_win !== 'function') {
```

**Review by brodrigu — CHANGES_REQUESTED** [review]
some small logic errors

**Review by brodrigu — APPROVED** [review]
☕

**Pull request merged**